implementation/index.html

```diff
@@ -0,0 +1,256 @@
+  console.assert("userChoice" in e1, assertion);
+} {
+  const assertion = "userChoice is a Promise";
+  const e1 = new BeforeInstallPromptEvent("");
+  console.assert(e1.userChoice instanceof Promise, assertion);
+} {
+  const assertion = "Correctly sets userChoice enum values";
+  const e1 = new BeforeInstallPromptEvent("", {
+    userChoice: "accepted"
+  });
+  e1.userChoice
+    .then(value => console.assert(value === "accepted"))
+    .catch(err => console.assert(false, assertion, err));
+  const e2 = new BeforeInstallPromptEvent("", {
+    userChoice: "dismissed"
+  });
+  e2.userChoice
+    .then(value => console.assert(value === "dismissed"))
+    .catch(err => console.assert(false, assertion, err));
+} {
+  const assertion = "BeforeInstallPromptEvent has a prompt() method";
+  const e = new BeforeInstallPromptEvent("");
+  console.assert("prompt" in e, assertion);
+} {
+  const assertion = "prompt() is a function";
+  const e = new BeforeInstallPromptEvent("");
+  console.assert(e.prompt instanceof Function, assertion);
+} {
+  const assertion = "prompt() takes no arguments";
+  const e = new BeforeInstallPromptEvent("");
+  console.assert(e.prompt.length === 0, assertion);
+} {
+  const assertion = "calling prompt with unstrusted event results in NotAllowedError error";
+  const e = new BeforeInstallPromptEvent("");
+  try {
+    e.prompt();
+    console.assert(false, assertion);
+  } catch (err) {
+    console.assert(err instanceof DOMException, assertion);
+    console.assert(err.type = "NotAllowedError", assertion);
+  }
+} {
+  const assertion = "prompt() first throws an NotAllowedError, then InvalidStateError";
+  const e = new BeforeInstallPromptEvent("");
+  try {
+    e.prompt();
+    console.assert(false, assertion);
+  } catch (err) {
+    console.assert(err instanceof DOMException, assertion);
+    console.assert(err.type = "NotAllowedError", assertion);
+  }
+  try {
+    e.prompt();
+    console.assert(false, assertion);
+  } catch (err) {
+    console.assert(err instanceof DOMException, assertion);
+    console.assert(err.type = "InvalidStateError", assertion);
+  }
+}
+
+// install events
+window.addEventListener("install", (ev) => {
+  console.info("Application installed successfully");
+});
+
+// Trusted events
+window.addEventListener("beforeinstallprompt", (ev) => {
+  {
+    const assertion = "only fires after document is fully loaded";
+    console.assert(document.readyState === "complete", assertion);
+  } {
+    const assertion = "is instance of BeforeInstallPromptEvent";
+    console.assert(ev instanceof BeforeInstallPromptEvent, assertion);
+  } {
+    const assertion = "is trusted";
+    console.assert(ev.isTrusted, assertion, ev);
+  } {
+    const assertion = "calling prompt successfully returns void";
+    console.assert(typeof ev.prompt() === "undefined", assertion);
+  } {
+    const assertion = "Even with success, prompt() can only be called once";
+    try {
+      ev.prompt();
```

Comment:
Why does this call `prompt` without first calling `preventDefault`?

I'm confused because you have another handler (in BIPE.js) that calls preventDefault and prompt. This seems to be testing that that event was handled first, before this one... but as we discussed previously the order isn't guaranteed. I think it would be best to have just a single bipe handler (in this file, not the other one) that tests all of these things.